drivers/i2c/i2c_omap.c

```diff
@@ -0,0 +1,947 @@
+/**
+ * @brief Get the state of the SCL line.
+ *
+ * This function retrieves the state of the SCL (clock) line for the OMAP I2C controller.
+ *
+ * @param io_context The I2C context.
+ * @return The state of the SCL line.
+ */
+static int __maybe_unused i2c_omap_get_scl(void *io_context)
+{
+	const struct i2c_omap_cfg *cfg = io_context;
+	uint32_t reg;
+
+	reg = i2c_omap_reg_access(cfg, I2C_SYSTEST);
+
+	return reg & I2C_SYSTEST_SCL_I_FUNC;
+}
+
+/**
+ * @brief Get the state of the SDA line.
+ *
+ * This function retrieves the state of the SDA (data) line for the OMAP I2C controller.
+ *
+ * @param io_context The I2C context.
+ * @return The state of the SDA line.
+ */
+static int i2c_omap_get_sda(void *io_context)
+{
+	const struct i2c_omap_cfg *cfg = io_context;
+	uint32_t reg;
+
+	reg = i2c_omap_reg_access(cfg, I2C_SYSTEST);
+
+	return reg & I2C_SYSTEST_SDA_I_FUNC;
+}
+
+/**
+ * @brief Set the state of the SDA line.
+ *
+ * This function sets the state of the SDA (data) line for the OMAP I2C controller.
+ *
+ * @param io_context The I2C context.
+ * @param state The state to set (0 for low, 1 for high).
+ */
+static void i2c_omap_set_sda(void *io_context, int state)
+{
+	const struct i2c_omap_cfg *cfg = io_context;
+	uint32_t reg;
+
+	reg = i2c_omap_reg_access(cfg, I2C_SYSTEST);
+
+	if (state) {
+		reg |= I2C_SYSTEST_SDA_O;
+	} else {
+		reg &= ~I2C_SYSTEST_SDA_O;
+	}
+
+	i2c_omap_reg_set(cfg, I2C_SYSTEST, reg);
+}
+
+/**
+ * @brief Set the state of the SCL line.
+ *
+ * This function sets the state of the SCL (clock) line for the OMAP I2C controller.
+ *
+ * @param io_context The I2C context.
+ * @param state The state to set (0 for low, 1 for high).
+ */
+static void i2c_omap_set_scl(void *io_context, int state)
+{
+	const struct i2c_omap_cfg *cfg = io_context;
+	uint32_t reg;
+
+	reg = i2c_omap_reg_access(cfg, I2C_SYSTEST);
+
+	if (state) {
+		reg |= I2C_SYSTEST_SCL_O;
+	} else {
+		reg &= ~I2C_SYSTEST_SCL_O;
+	}
+
+	i2c_omap_reg_set(cfg, I2C_SYSTEST, reg);
+}
+/**
+ * @brief Recovers the I2C bus using the OMAP I2C controller.
+ *
+ * This function attempts to recover the I2C bus by performing a bus recovery
+ * sequence using the OMAP I2C controller. It uses the provided device
+ * configuration and bit-banging operations to recover the bus.
+ *
+ * @param dev Pointer to the device structure.
+ * @return 0 on success, negative error code on failure.
+ */
+
+static int i2c_omap_recover_bus(const struct device *dev)
+{
+	const struct i2c_omap_cfg *cfg = dev->config;
+	struct i2c_bitbang bitbang_omap;
+	struct i2c_bitbang_io bitbang_omap_io = {
+		.get_sda = i2c_omap_get_sda,
+		.set_scl = i2c_omap_set_scl,
+		.set_sda = i2c_omap_set_sda,
+	};
+	uint32_t reg;
+	int error = 0;
+
+	reg = i2c_omap_reg_access(cfg, I2C_SYSTEST);
+	reg |= I2C_SYSTEST_ST_EN | 3 << I2C_SYSTEST_TMODE_SHIFT | I2C_SYSTEST_SCL_O |
+	       I2C_SYSTEST_SDA_O;
+	i2c_omap_reg_set(cfg, I2C_SYSTEST, reg);
+
+	i2c_bitbang_init(&bitbang_omap, &bitbang_omap_io, (void *)cfg);
+
```

Comment:
Can you explain what is being done here? who cleans up after recovery? can you explain in the commit message the details of the implementation of the driver atleast wherever it deviates from standard implementations like Linux/U-Boot/TI SDKs?

Reply:
I based out the recovery from [i2c_bitbang_recover_bus](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_bitbang.c#L277), which enables bit bang operations to send 9 SCL clk pulses.

Post that, it checks if the line is down. and then disables the system_test register and . resuming the transaction 

Will update this in the commit message

Reply:
Done!